velox/functions/prestosql/window/FirstLastValue.cpp

```diff
@@ -0,0 +1,134 @@
+ public:
+  explicit FirstLastValueFunction(
+      const std::vector<exec::WindowFunctionArg>& args,
+      const TypePtr& resultType,
+      velox::memory::MemoryPool* pool)
+      : WindowFunction(resultType, pool, nullptr) {
+    VELOX_CHECK_NULL(args[0].constantValue);
+    argIndex_ = args[0].index.value();
+  }
+
+  void resetPartition(const exec::WindowPartition* partition) override {
+    partition_ = partition;
+  }
+
+  void apply(
+      const BufferPtr& /*peerGroupStarts*/,
+      const BufferPtr& /*peerGroupEnds*/,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds,
+      const SelectivityVector& validRows,
+      int32_t resultOffset,
+      const VectorPtr& result) override {
+    auto numRows = frameStarts->size() / sizeof(vector_size_t);
+    rowNumbers_.resize(numRows);
+
+    if constexpr (TValue == ValueType::kFirst) {
+      auto rawFrameStarts = frameStarts->as<vector_size_t>();
+      validRows.applyToSelected(
+          [&](auto i) { rowNumbers_[i] = rawFrameStarts[i]; });
+    } else {
+      auto rawFrameEnds = frameEnds->as<vector_size_t>();
+      validRows.applyToSelected(
+          [&](auto i) { rowNumbers_[i] = rawFrameEnds[i]; });
+    }
+    setRowNumbersForEmptyFrames(validRows);
+
+    auto rowNumbersRange = folly::Range(rowNumbers_.data(), numRows);
+    partition_->extractColumn(argIndex_, rowNumbersRange, resultOffset, result);
+  }
+
+ private:
+  void setRowNumbersForEmptyFrames(const SelectivityVector& validRows) {
+    if (validRows.isAllSelected()) {
+      return;
+    }
+    // Rows with empty (not-valid) frames have nullptr in the result.
+    // So mark rowNumber to copy as -1 for it.
+    invalidRows_.resizeFill(validRows.size(), true);
+    invalidRows_.deselect(validRows);
+    invalidRows_.applyToSelected([&](auto i) { rowNumbers_[i] = -1; });
+  }
+
+  // Index of the first_value / last_value argument column in the input row
+  // vector. This is used to retrieve column values from the partition data.
+  column_index_t argIndex_;
+
+  const exec::WindowPartition* partition_;
+
+  // The first_value, last_value functions directly write from the input column
+  // to the resultVector using the extractColumn API specifying the rowNumber
+  // mapping to copy between the 2 vectors. This variable is used for the
+  // rowNumber vector across getOutput calls.
+  std::vector<vector_size_t> rowNumbers_;
+
+  // Member variable re-used for setting null for empty frames.
+  SelectivityVector invalidRows_;
+};
+} // namespace
+
+template <ValueType TValue>
+void registerFirstLastInternal(const std::string& name) {
+  // first_value(T) -> T
+  // last_value(T) -> T
+  std::vector<exec::FunctionSignaturePtr> signatures{
+      exec::FunctionSignatureBuilder()
+          .typeVariable("T")
+          .returnType("T")
+          .argumentType("T")
+          .build(),
+  };
+
+  exec::registerWindowFunction(
+      name,
+      std::move(signatures),
+      [name](
```

Comment:
name is not used; let's remove

Reply:
Thanks for merging the changes. Apologies for the delay in addressing the pending comments, I was analyzing the benchmarks and working on reducing the test time. Will make these changes in a subsequent PR.